File: job-finder-worker/src/job_finder/ai/agent_manager.py
"""Agent Manager for AI provider selection and lifecycle management.

The AgentManager abstracts AI agent selection from callers. Workers supply a task type
and prompt; the manager selects the appropriate agent based on configuration,
availability, and budget.

Key responsibilities:
- Select agents from fallback chain based on task type
- Enforce daily budget limits before calling agents
- Disable agents on errors (quota exhaustion, API failures)
- Track usage via config updates
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Any, Dict, Optional, TYPE_CHECKING

from job_finder.ai.providers import auth_status

from job_finder.exceptions import AIProviderError, NoAgentsAvailableError, QuotaExhaustedError

if TYPE_CHECKING:
    from job_finder.job_queue.config_loader import ConfigLoader

logger = logging.getLogger(__name__)


@dataclass
class AgentResult:
    """Result from an agent execution."""

    text: str
    agent_id: str
    model: str


# Provider dispatch map: (provider, interface) -> provider class
# Import lazily to avoid circular imports
def _get_provider_class(provider: str, interface: str):
    """Get the provider class for a provider/interface combination."""
    from job_finder.ai.providers import (
        ClaudeProvider,
        CodexCLIProvider,
        GeminiCLIProvider,
        GeminiProvider,
        OpenAIProvider,
    )

    provider_map = {
        ("codex", "cli"): CodexCLIProvider,
        ("claude", "api"): ClaudeProvider,
        ("openai", "api"): OpenAIProvider,
        ("gemini", "api"): GeminiProvider,
        ("gemini", "cli"): GeminiCLIProvider,
    }
    return provider_map.get((provider, interface))


class AgentManager:
    """Manages AI agent selection, fallback, and lifecycle.

    The manager reads ai-settings fresh on every call to support shared budget
    tracking between worker and backend. It enforces budget limits before calling
    agents and disables agents on errors.
    """

    def __init__(self, config_loader: "ConfigLoader"):
        """Initialize the agent manager.

        Args:
            config_loader: ConfigLoader instance for reading/writing ai-settings
        """
        self.config_loader = config_loader

    def execute(
        self,
        task_type: str,
        prompt: str,
        model_override: Optional[str] = None,
        max_tokens: int = 1000,
        temperature: float = 0.7,
    ) -> AgentResult:
        """Execute a task using the appropriate agent from the fallback chain.

        Tries agents in fallback order until one succeeds. Handles budget
        enforcement and error-based agent disabling.

        Args:
            task_type: The type of task ("extraction" or "analysis")
            prompt: The prompt to send to the agent
            model_override: Optional model to use instead of agent default
            max_tokens: Maximum tokens in response
            temperature: Sampling temperature

        Returns:
            AgentResult with response text and metadata

        Raises:
            NoAgentsAvailableError: All agents exhausted, disabled, or over budget
        """
        # Always read fresh config (enables shared budget with backend)
        ai_settings = self.config_loader.get_ai_settings()

        fallback_chain = ai_settings.get("taskFallbacks", {}).get(task_type, [])
        if not fallback_chain:
            raise NoAgentsAvailableError(
                f"No fallback chain configured for task type: {task_type}",
                task_type=task_type,
            )

        agents = ai_settings.get("agents", {})
        model_rates = ai_settings.get("modelRates", {})
        tried_agents = []
        errors = []

        for agent_id in fallback_chain:
            agent_config = agents.get(agent_id)
            if not agent_config:
                logger.warning(f"Agent {agent_id} in fallback chain not found in agents config")
                continue

            tried_agents.append(agent_id)

            # Skip disabled agents
            if not agent_config.get("enabled", True):
                reason = agent_config.get("reason", "unknown")
                logger.debug(f"Skipping disabled agent {agent_id}: {reason}")
                continue

            provider = agent_config.get("provider")
            interface = agent_config.get("interface")
            if not provider or not interface:
                logger.warning(f"Agent {agent_id} missing provider/interface; skipping")
                continue

            auth_ok, auth_reason = auth_status(provider, interface)
            if not auth_ok:
                agent_config["enabled"] = False
                agent_config["reason"] = auth_reason
                try:
                    self.config_loader.update_agent_status(agent_id, enabled=False, reason=auth_reason)
                except Exception as exc:  # best-effort, don't crash caller
                    logger.warning(f"Failed to persist disable for {agent_id}: {exc}")
                logger.info(f"Skipping agent {agent_id}: {auth_reason}")
                continue

            # Determine model first so we can calculate cost for budget check
            model = model_override or agent_config.get("defaultModel")
            cost = model_rates.get(model, 1.0)

            # Budget enforcement - check before calling (using model cost)
            daily_usage = agent_config.get("dailyUsage", 0)
            daily_budget = agent_config.get("dailyBudget", float("inf"))
            if daily_usage + cost > daily_budget:
                logger.info(
                    f"Agent {agent_id} over budget ({daily_usage}+{cost}/{daily_budget}), disabling"
                )
                self._disable_agent(agent_id, "quota_exhausted: daily budget reached")
                continue

            # Try to call the agent
            try:
                result = self._call_agent(
                    agent_id, agent_config, prompt, model, max_tokens, temperature
                )

                # Increment usage after successful call
                cost = model_rates.get(model, 1.0)
                self.config_loader.increment_agent_usage(agent_id, model)
                logger.info(f"Agent {agent_id} executed successfully (model={model}, cost={cost})")

                return result

            except QuotaExhaustedError as e:
                # Quota/rate limit errors - disable agent but continue to next
                error_msg = str(e)
                logger.warning(f"Agent {agent_id} quota exhausted: {error_msg}")
                self._disable_agent(agent_id, f"quota_exhausted: {error_msg}")
                errors.append((agent_id, error_msg))
                continue  # Try next agent in fallback chain

            except AIProviderError as e:
                # Other API errors - disable agent and stop (requires investigation)
                error_msg = str(e)
                logger.error(f"Agent {agent_id} failed: {error_msg}")
                self._disable_agent(agent_id, f"error: {error_msg}")
                errors.append((agent_id, error_msg))
                # Don't continue - non-quota errors may indicate systemic issues
                break

        # All agents exhausted
        error_summary = (
            "; ".join(f"{aid}: {err}" for aid, err in errors)
            if errors
            else "all disabled/over budget"
        )
        raise NoAgentsAvailableError(
            f"No agents available for task '{task_type}'. Tried: {tried_agents}. Errors: {error_summary}",
            task_type=task_type,
            tried_agents=tried_agents,
        )

    def _call_agent(
        self,
        agent_id: str,
        agent_config: Dict[str, Any],
        prompt: str,
        model: str,
        max_tokens: int,
        temperature: float,
    ) -> AgentResult:
        """Instantiate and call the appropriate provider.

        Args:
            agent_id: The agent ID for logging
            agent_config: Agent configuration dict
            prompt: The prompt to send
            model: Model to use
            max_tokens: Max tokens in response
            temperature: Sampling temperature

        Returns:
            AgentResult with response text
        """
        provider_type = agent_config.get("provider")
        interface_type = agent_config.get("interface")

        provider_class = _get_provider_class(provider_type, interface_type)
        if not provider_class:
            raise AIProviderError(f"No provider class for {provider_type}/{interface_type}")

        logger.debug(f"Calling {agent_id} with model={model}")
        provider = provider_class(model=model)
        response = provider.generate(prompt, max_tokens=max_tokens, temperature=temperature)

        return AgentResult(text=response, agent_id=agent_id, model=model)

    def _disable_agent(self, agent_id: str, reason: str) -> None:
        """Disable an agent with a reason.

        Args:
            agent_id: The agent ID to disable
            reason: Why the agent is being disabled
        """
        self.config_loader.update_agent_status(agent_id, enabled=False, reason=reason)
        logger.warning(f"Disabled agent {agent_id}: {reason}")

File: job-finder-BE/server/src/modules/generator/ai/agent-manager.ts
import { existsSync } from 'node:fs'
import { homedir } from 'node:os'
import { join } from 'node:path'
import type { Logger } from 'pino'
import type { AISettings, AgentConfig, AgentId, AgentScope, AgentTaskType } from '@shared/types'
import { ConfigRepository } from '../../config/config.repository'
import { logger } from '../../../logger'
import { runCliProvider, type CliProvider, type CliErrorType } from '../workflow/services/cli-runner'
import { UserFacingError } from '../workflow/generator.workflow.service'

type AgentExecutionResult = {
  output: string
  agentId: string
  model: string | undefined
}

class NoAgentsAvailableError extends Error {
  constructor(message: string, readonly taskType: AgentTaskType, readonly triedAgents: string[]) {
    super(message)
  }
}

/** Errors that should continue to the next agent in the fallback chain */
class QuotaExhaustedError extends Error {
  constructor(message: string, readonly agentId: string) {
    super(message)
  }
}

/** Errors that should stop the fallback chain (systemic issues) */
class AgentExecutionError extends Error {
  constructor(message: string, readonly agentId: string, readonly errorType: CliErrorType) {
    super(message)
  }
}

function expandHome(filePath: string): string {
  if (!filePath.startsWith('~')) return filePath
  return join(homedir(), filePath.slice(1))
}

export class AgentManager {
  private readonly scope: AgentScope = 'backend'

  constructor(private readonly configRepo = new ConfigRepository(), private readonly log: Logger = logger) {}

  ensureAvailable(taskType: AgentTaskType): void {
    const entry = this.configRepo.get<AISettings>('ai-settings')
    if (!entry?.payload) {
      throw new UserFacingError('AI settings not configured. Please configure ai-settings in the database.')
    }
    const chain = entry.payload.taskFallbacks?.[taskType]
    if (!Array.isArray(chain) || chain.length === 0) {
      throw new UserFacingError(`No fallback chain configured for task ${taskType}`)
    }

    const agents = entry.payload.agents
    const hasEnabled = chain.some((agentId) => agents?.[agentId]?.runtimeState?.[this.scope]?.enabled)
    if (!hasEnabled) {
      throw new UserFacingError(`No enabled agents for task ${taskType} in scope ${this.scope}`)
    }
  }

  async execute(taskType: AgentTaskType, prompt: string, modelOverride?: string): Promise<AgentExecutionResult> {
    const entry = this.configRepo.get<AISettings>('ai-settings')
    if (!entry?.payload) {
      throw new UserFacingError('AI settings not configured. Please configure ai-settings in the database.')
    }

    const aiSettings = entry.payload
    const chain = aiSettings.taskFallbacks?.[taskType]

    if (!Array.isArray(chain) || chain.length === 0) {
      throw new NoAgentsAvailableError(`No fallback chain configured for ${taskType}`, taskType, [])
    }

    const tried: string[] = []

    for (const agentId of chain) {
      const agent = aiSettings.agents?.[agentId]
      if (!agent) {
        this.log.warn({ agentId }, 'Agent in fallback chain not found')
        continue
      }
      tried.push(agentId)

      const scopeState = agent.runtimeState?.[this.scope]
      if (!scopeState) {
        throw new NoAgentsAvailableError(`Missing runtimeState for scope ${this.scope}`, taskType, tried)
      }
      if (!scopeState.enabled) {
        this.log.debug({ agentId, reason: scopeState.reason }, 'Skipping disabled agent for backend scope')
        continue
      }

      const authError = this.checkAuth(agent)
      if (authError) {
        this.disableAgent(aiSettings, agentId, authError)
        continue
      }

      // Determine model - "default" or empty means CLI uses its own default
      const rawModel = modelOverride ?? agent.defaultModel
      const model = rawModel && rawModel !== 'default' ? rawModel : undefined
      // Cost defaults to 1.0 when model is undefined (CLI using its default)
      const cost = model ? (aiSettings.modelRates?.[model] ?? 1) : 1
      if (agent.dailyUsage + cost > agent.dailyBudget) {
        this.disableAgent(aiSettings, agentId, 'quota_exhausted: daily budget reached')
        continue
      }

      try {
        const output = await this.runAgent(agent, agentId, prompt, model)
        agent.dailyUsage += cost
        this.persist(aiSettings)
        return { output, agentId, model }
      } catch (err) {
        if (err instanceof QuotaExhaustedError) {
          // Quota errors: disable agent and continue to next in fallback chain
          this.log.warn({ agentId, error: err.message }, 'Agent quota exhausted, trying next agent')
          this.disableAgent(aiSettings, agentId, `quota_exhausted: ${err.message}`)
          continue
        }

        if (err instanceof AgentExecutionError) {
          // Non-quota errors: disable agent and throw immediately (systemic issues)
          this.log.error({ agentId, error: err.message, errorType: err.errorType }, 'Agent execution failed')
          this.disableAgent(aiSettings, agentId, `error: ${err.message}`)
          throw new UserFacingError(`AI generation failed: ${err.message}`)
        }

        // Unknown errors: disable and re-throw
        const reason = err instanceof Error ? err.message : 'agent failed'
        this.disableAgent(aiSettings, agentId, `error: ${reason}`)
        throw err
      }
    }

    throw new NoAgentsAvailableError(
      `No agents available for task ${taskType}. Tried: ${tried.join(', ')}`,
      taskType,
      tried
    )
  }

  private checkAuth(agent: AgentConfig): string | null {
    const { authRequirements } = agent
    const envList = authRequirements.requiredEnv ?? []
    const fileList = authRequirements.requiredFiles ?? []

    const envSatisfied = envList.length === 0 ? false : envList.some((env) => !!process.env[env])
    const fileSatisfied = fileList.length === 0 ? false : fileList.some((file) => existsSync(expandHome(file)))

    const authOk = envList.length && fileList.length ? envSatisfied || fileSatisfied : envList.length ? envSatisfied : fileList.length ? fileSatisfied : true

    if (authOk) return null

    const parts: string[] = []
    if (envList.length) parts.push(`missing_env:any_of:${envList.join(',')}`)
    if (fileList.length) parts.push(`missing_file:any_of:${fileList.join(',')}`)
    return parts.join('|')
  }

  private async runAgent(agent: AgentConfig, agentId: string, prompt: string, model: string | undefined): Promise<string> {
    if (agent.interface !== 'cli') {
      throw new UserFacingError(`Interface ${agent.interface} not supported for generator tasks`)
    }
    const provider = agent.provider as CliProvider
    const result = await runCliProvider(prompt, provider, { model })

    if (!result.success) {
      const errorMsg = result.error || 'AI generation failed'
      const errorType = result.errorType || 'other'

      // Throw appropriate error type based on CLI result
      if (errorType === 'quota') {
        throw new QuotaExhaustedError(errorMsg, agentId)
      }

      // Auth, timeout, not_found, and other errors are considered systemic
      throw new AgentExecutionError(errorMsg, agentId, errorType)
    }

    return result.output
  }

  private disableAgent(aiSettings: AISettings, agentId: AgentId, reason: string): void {
    const agent = aiSettings.agents?.[agentId]
    if (!agent) {
      throw new UserFacingError(`Agent ${agentId} missing from ai-settings`)
    }
    agent.runtimeState[this.scope] = { enabled: false, reason }
    this.persist(aiSettings)
    this.log.warn({ agentId, scope: this.scope, reason }, 'Disabled agent')
  }

  private persist(aiSettings: AISettings): void {
    this.configRepo.upsert('ai-settings', aiSettings, { updatedBy: 'agent-manager-backend' })
  }
}

export { NoAgentsAvailableError, QuotaExhaustedError, AgentExecutionError, AgentExecutionResult }
